### PR TITLE
[fix]thrift_proxy:Fixed potential assert failure when input contains invalid characters

### LINKED ISSUE #12361: [api] cluster_header in thrift_proxy needs validation
Assert failure will occur when instantiate an thrift_proxy network-level filter while `config.route_config.route.cluster_header` contains invalid characters for http header.

[assert] [bazel-out/k8-fastbuild/bin/include/envoy/http/_virtual_includes/header_map_interface/envoy/http/header_map.h:54] assert failure: valid().

Need to add a validation `well_known_regex: HTTP_HEADER_VALUE strict: false` to that field.

### PR DESCRIPTION
**Pull request by jianwen612**
The assert failure will occur when `config.route_config.route.cluster_header` contains invalid characters for the HTTP header. Added the validation for this field to avoid assert failure and regression a test case in the unit test.

```
[assert] [bazel-out/k8-fastbuild/bin/include/envoy/http/_virtual_includes/header_map_interface/envoy/http/header_map.h:54] assert failure: valid().
```

Risk Level: Low
Testing: Added a regression test case
Fixes #12361

/cc @asraa 
/cc @samkerner

### CODE REVIEW

**Comment by jianwen612** [comment]
@zuercher 
Thanks for pointing out that there are several other places where we need to remove `\0\r\n` characters. 
I added removing operations to 4 places in `twitter_protocol_impl.cc` and `header_transport_impl.cc`.
If there are any other places which we need to modify, please let me know!

File: source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc [hunk 1]
```diff
@@ -1022,7 +1024,10 @@ void TwitterProtocolImpl::updateMetadataWithRequestHeader(const ThriftObject& he
     metadata.setFlags(*req_header.flags());
   }
   for (const auto& context : *req_header.contexts()) {
-    headers.addCopy(Http::LowerCaseString{context.key_}, context.value_);
+    // LowerCaseString doesn't allow '\0', '\n', and '\r'.
+    std::string key =
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc:1028]: nit: `const`

Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc:1028]: nit: `const`

File: source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc [hunk 2]
```diff
@@ -1031,6 +1036,8 @@ void TwitterProtocolImpl::updateMetadataWithRequestHeader(const ThriftObject& he
   // objects
   for (const auto& delegation : *req_header.delegations()) {
     std::string key = fmt::format(":d:{}", delegation.src_);
+    // LowerCaseString doesn't allow '\0', '\n', and '\r'.
+    key = absl::StrReplaceAll(key, {{std::string(1, '\0'), ""}, {"\n", ""}, {"\r", ""}});
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc:1040]: nit: `const std::string key` and then just call `absl::StrReplaceAll()` on `delegation.src_`

File: source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc [hunk 3]
```diff
@@ -1050,7 +1057,10 @@ void TwitterProtocolImpl::updateMetadataWithResponseHeader(const ThriftObject& h
 
   Http::HeaderMap& headers = metadata.headers();
   for (const auto& context : resp_header.contexts()) {
-    headers.addCopy(Http::LowerCaseString(context.key_), context.value_);
+    // LowerCaseString doesn't allow '\0', '\n', and '\r'.
+    std::string key =
+        absl::StrReplaceAll(context.key_, {{std::string(1, '\0'), ""}, {"\n", ""}, {"\r", ""}});
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/twitter_protocol_impl.cc:1062]: nit `const`

**Comment by jianwen612** [comment]
> Thanks! Just left a couple of nits (feel free to ignore them).

Thanks for pointing those out! Fixed!

**Comment by zuercher** [comment]
Forgot one thing -- since we're changing the behavior here we need to add a release note saying that those special characters will be stripped from thrift headers. I'm going to be away from computers next week, but I'm ok if another maintainer wants to verify the addition of the release note and merge.

**Comment by asraa** [comment]
> I'm going to be away from computers next week, but I'm ok if another maintainer wants to verify the addition of the release note and merge.

Good idea, will do! 
/wait

**Comment by jianwen612** [comment]
> > I'm going to be away from computers next week, but I'm ok if another maintainer wants to verify the addition of the release note and merge.
> 
> Good idea, will do!
> /wait

Added!

**Pull request merged**